### ISSUE TITLE
FLAT v IVF FLAT Divergence

### ISSUE THREAD

**Issue by CatEggs** [milvus-io/pymilvus#485]
**Describe the bug**

_Created a test to understand where IVF Flat and FLAT index search results diverge. 

Using scikit learns make_blobs, I inserted 1000 samples into 2 different milvus collection with center being  a random set of numbers of size (10,7). 
I set nlist =2 and nprobe=1.
When I attempt to query these the collections, I get the same out results from the IVF FLAT index as I do the FLAT index. except when I modify my center size to (10,8) and my search dimension=8. To understand the behavior of IVF FLAT, I would like some clarity as to why IVF FLAT fails to partition when the dimension is less than 8._

**Steps/Code to reproduce behavior**
Follow this [guide](http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) to craft a minimal bug report. This helps us reproduce the issue you're having and resolve the issue more quickly.

 -------
from milvus import Milvus, IndexType, MetricType
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import make_blobs
import numpy as np
import random

centers = np.random.uniform(size=(10, 7))
X, labels_true = make_blobs(n_samples=1000, centers=centers, cluster_std=0.1,
                            random_state=0)

X = StandardScaler().fit_transform(X)
## FLAT
param = {'collection_name':'test_milvus', 'dimension':7, 'index_file_size':1024, 'metric_type':MetricType.L2}

milvus.create_collection(param)

from pprint import pprint

vector_ids = [id for id in range(1000)]

status, inserted_vector_ids = milvus.insert(collection_name='test_milvus', records=X.tolist(), ids = vector_ids)

print(status)

#IVF FLAT
param_2 = {'collection_name':'test_milvus_ivf', 'dimension':7, 'index_file_size':1024, 'metric_type':MetricType.L2}

milvus.create_collection(param_2)

ivf_param = {'nlist': 2}

milvus.create_index('test_milvus_ivf', IndexType.IVF_FLAT, ivf_param)

status, inserted_vector_ids = milvus.insert(collection_name='test_milvus_ivf', records=X.tolist(), ids = vector_ids)

print(status)

#Should not need this since FLAT is the default but just in case ##

milvus.create_index('test_milvus', IndexType.FLAT)

search_param = {'nprobe': 1}

#Create test search query 

q_records = [[random.random() for _ in range(7)] for _ in range(100)]

status, results_ivf = milvus.search(collection_name='test_milvus_ivf', query_records=q_records, top_k=100, params=search_param)

status, results_flat = milvus.search(collection_name='test_milvus', query_records=q_records, top_k=100, params={})

#Test search result output 

same_ids=[]
diff_ids=[]
for i in range(len(results_flat)):
    for j in range(100):
        if (results_flat[i][j].distance == results_ivf[i][j].distance):
            same_ids.append({"flat_ids":results_flat[i][j].id, "ivf_ids":results_ivf[i][j].id,
                            "flat_distance":results_flat[i][j].distance, "ivf_distance":results_ivf[i][j].distance})
        else:
            diff_ids.append({"flat_ids":results_flat[i][j].id, "ivf_ids":results_ivf[i][j].id,
                            "flat_distance":results_flat[i][j].distance, "ivf_distance":results_ivf[i][j].distance})
print(len(diff_ids))

**Expected behavior**

_Expected: The length of the diff_ids not  equal to 0._

**Environment details**

- Hardware/Softward conditions (OS, CPU, GPU, Memory)
- Method of installation (Docker, or from source)
- Milvus version (v0.3.1, or v0.4.0)
- Milvus configuration (Settings you made in `server_config.yaml`)

**Screenshots**

**Additional context**

**Comment by XuanYang-cn** [comment]
https://github.com/milvus-io/pymilvus/issues/485#issue-861799311 better-looking codes
```python
from milvus import Milvus, IndexType, MetricType
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import make_blobs
import numpy as np
import random

centers = np.random.uniform(size=(10, 7))
X, labels_true = make_blobs(n_samples=1000, centers=centers, cluster_std=0.1,
random_state=0)

X = StandardScaler().fit_transform(X)

# FLAT
param = {'collection_name':'test_milvus', 'dimension':7, 'index_file_size':1024, 'metric_type':MetricType.L2}
milvus.create_collection(param)
from pprint import pprint

vector_ids = [id for id in range(1000)]
status, inserted_vector_ids = milvus.insert(collection_name='test_milvus', records=X.tolist(), ids = vector_ids)
print(status)

# IVF FLAT
param_2 = {'collection_name':'test_milvus_ivf', 'dimension':7, 'index_file_size':1024, 'metric_type':MetricType.L2}
milvus.create_collection(param_2)
ivf_param = {'nlist': 2}
milvus.create_index('test_milvus_ivf', IndexType.IVF_FLAT, ivf_param)
status, inserted_vector_ids = milvus.insert(collection_name='test_milvus_ivf', records=X.tolist(), ids = vector_ids)
print(status)

# Should not need this since FLAT is the default but just in case ##

milvus.create_index('test_milvus', IndexType.FLAT)
search_param = {'nprobe': 1}

# Create test search query
q_records = [[random.random() for _ in range(7)] for _ in range(100)]
status, results_ivf = milvus.search(collection_name='test_milvus_ivf', query_records=q_records, top_k=100, params=search_param)
status, results_flat = milvus.search(collection_name='test_milvus', query_records=q_records, top_k=100, params={})

#Test search result output
same_ids=[]
diff_ids=[]
for i in range(len(results_flat)):
    for j in range(100):
        if (results_flat[i][j].distance == results_ivf[i][j].distance):
            same_ids.append({"flat_ids":results_flat[i][j].id, "ivf_ids":results_ivf[i][j].id,
                "flat_distance":results_flat[i][j].distance, "ivf_distance":results_ivf[i][j].distance})
        else:
            diff_ids.append({"flat_ids":results_flat[i][j].id, "ivf_ids":results_ivf[i][j].id,
                "flat_distance":results_flat[i][j].distance, "ivf_distance":results_ivf[i][j].distance})
print(len(diff_ids))
```

**Comment by shengjun1985** [comment]
Please refer to: [Create-indexes](https://milvus.io/docs/v1.0.0/index.md#Create-indexes)
When the inserted data segment is less than **4096** rows, Milvus does not index it.

You inserted too little data, so the index didn't work.

**Comment by shengjun1985** [comment]
index = IVF_FLAT, nlist = 2, nprobe = 1
Only when you query the points whose distance from the two centroids are almost the same, the results will be different from that of FLAT. Therefor, the query results are greatly affected by the data distribution. 
Are there such points in your query?

**Comment by CatEggs** [comment]
Hi there! Thanks again for the help. When you say "almost the same" would you happen to know exactly how far or close the points need to be in order to get that divergence? I'm writing a function script to check do a check

**Comment by CatEggs** [comment]
Hi there,

So I modified my code slightly. I readjusted so that there should only produce 1 cluster and increased the sample size to 4097. I kept the nlist=2 and the nprobe=1. Same as before I am comparing the search results from IVF Flat and Flat index to diverge but I still get the same result. Please let me know if I should make changes to my test to identify where the search results diverge. 
___________________________________________________________________________________________________________________________________

from milvus import Milvus, IndexType, MetricType
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import make_blobs
import numpy as np
import random

np.random.seed(0)
center_count =1
dim = 2
n = 4097*center_count
nlist = 2
nprobe = 1

centers = np.random.uniform(size=(center_count, dim)).astype('float32')
X, labels_true = make_blobs(n_samples=n, centers=centers, cluster_std=0.1,
                            random_state=0)

X = StandardScaler().fit_transform(X).astype('float32')

milvus = Milvus()
milvus.drop_collection('test_milvus_ivf')
milvus.drop_collection('test_milvus_flat')

### FLAT Index ###
param = {'collection_name':'test_milvus_flat', 'dimension':dim, 'index_file_size':32, 'metric_type':MetricType.L2}

milvus.create_collection(param)

from pprint import pprint

vector_ids = [id for id in range(n)]
vector_ids_ivf = [id for id in range(n+1,n+n+1)]

status, inserted_vector_ids = milvus.insert(collection_name='test_milvus_flat', records=X.tolist(), ids = vector_ids)
milvus.flush(['test_milvus_flat'])
print(status)

### IVF_FLAT Index ###
param_2 = {'collection_name':'test_milvus_ivf', 'dimension':dim, 'index_file_size':32, 'metric_type':MetricType.L2}

milvus.create_collection(param_2)

ivf_param = {'nlist': nlist}

milvus.create_index('test_milvus_ivf', IndexType.IVF_FLAT, ivf_param)

status, inserted_vector_ids = milvus.insert(collection_name='test_milvus_ivf', records=X.tolist(), ids = vector_ids_ivf)

milvus.flush(['test_milvus_ivf'])
print(status)


### Create test search query# ##
### Should not need this since FLAT is the default but just in case ###

milvus.create_index('test_milvus_flat', IndexType.FLAT)

search_param = {'nprobe': nprobe}
q_records = [[random.random() for _ in range(dim)] for _ in range(1000)]

status, results_ivf = milvus.search(collection_name='test_milvus_ivf', query_records=q_records, top_k=5, params=search_param)

status, results_flat = milvus.search(collection_name='test_milvus_flat', query_records=q_records, top_k=5, params={})

### Test Search Results ###
same_ids=[]
diff_ids=[]
for i in range(len(results_ivf)):
    for j in range(5):
        if (results_flat[i][j].distance == results_ivf[i][j].distance):
            same_ids.append({"flat_ids":results_flat[i][j].id, "ivf_ids":results_ivf[i][j].id,
                            "flat_distance":results_flat[i][j].distance, "ivf_distance":results_ivf[i][j].distance})
        else:
            diff_ids.append({"flat_ids":results_flat[i][j].id, "ivf_ids":results_ivf[i][j].id,
                            "flat_distance":results_flat[i][j].distance, "ivf_distance":results_ivf[i][j].distance})

print((len(diff_ids)))

**Comment by XuanYang-cn** [comment]
@shengjun1985

**Comment by shengjun1985** [comment]
I am sorry that I didn't find another problem in your code last time.

```python
milvus.create_index('test_milvus_ivf', IndexType.IVF_FLAT, ivf_param)
status, inserted_vector_ids = milvus.insert(collection_name='test_milvus_ivf', records=X.tolist(), ids = vector_ids_ivf)
milvus.flush(['test_milvus_ivf'])
```
should be:
```python
status, inserted_vector_ids = milvus.insert(collection_name='test_milvus_ivf', records=X.tolist(), ids = vector_ids_ivf)
milvus.flush(['test_milvus_ivf'])
milvus.create_index('test_milvus_ivf', IndexType.IVF_FLAT, ivf_param)
```

When the create_index method is called, Milvus **synchronously** indexes the **existing** data on this field.
Whenever the size of the inserted data reaches the index_file_size, Milvus automatically creates an index for it in the background.

**Comment by CatEggs** [comment]
That was my issue, thank you!
